File: DevExpress-Demo/VB/PivotGridDemo.Wpf/ModuleResources/CustomerReportsViewModel.vb
Imports System.Collections.Generic
Imports System.Linq
Imports DevExpress.DemoData
Imports DevExpress.Mvvm.DataAnnotations
Imports DevExpress.Mvvm.Native

Namespace PivotGridDemo

    <POCOViewModel>
    Public Class CustomerReportsViewModel

        Const allString As String = "(All)"

        Public Sub New()
            AllYears = CustomerReports.[Select](Function(x) x.OrderDate.Value.Year).Distinct().OrderBy(Function(x) x).Cast(Of Object)()
            AllQuarters = New Object() {allString}.Concat(CustomerReports.[Select](Function(x)(x.OrderDate.Value.Month - 1) \ 3 + 1).Distinct().OrderBy(Function(x) x).Cast(Of Object)())
            SelectedQuarter = AllQuarters.FirstOrDefault()
            SelectedYear = AllYears.FirstOrDefault()
        End Sub

        Public Property AllQuarters As IEnumerable(Of Object)

        Public Property AllYears As IEnumerable(Of Object)

        Public ReadOnly Property CustomerReports As IList(Of Models.CustomerReport)
            Get
                Return NWindDataProvider.CustomerReports
            End Get
        End Property

        Public Overridable Property QuartersFilter As Object

        <BindableProperty(OnPropertyChangedMethodName:="OnSelectedQuarterChanged")>
        Public Overridable Property SelectedQuarter As Object

        <BindableProperty(OnPropertyChangedMethodName:="OnSelectedYearChanged")>
        Public Overridable Property SelectedYear As Object

        Public Overridable Property YearsFilter As Object

        Protected Sub OnSelectedQuarterChanged()
            QuartersFilter = If(Equals(SelectedQuarter, allString), AllQuarters.Skip(1), {SelectedQuarter})
        End Sub

        Protected Sub OnSelectedYearChanged()
            YearsFilter = {SelectedYear}
        End Sub
    End Class
End Namespace

File: DevExpress-Demo/CS/PivotGridDemo.Wpf/ModuleResources/CustomerReportsViewModel.cs
using System.Collections.Generic;
using System.Linq;
using DevExpress.DemoData;
using DevExpress.Mvvm.DataAnnotations;
using DevExpress.Mvvm.Native;

namespace PivotGridDemo {
    [POCOViewModel]
    public class CustomerReportsViewModel {
        const string allString = "(All)";

        public CustomerReportsViewModel() {
            AllYears = CustomerReports
                .Select(x => x.OrderDate.Value.Year)
                .Distinct()
                .OrderBy(x => x)
                .Cast<object>();
            AllQuarters = new object[] { allString }.Concat(
                CustomerReports
                .Select(x => (x.OrderDate.Value.Month - 1) / 3 + 1)
                .Distinct()
                .OrderBy(x => x)
                .Cast<object>());

            SelectedQuarter = AllQuarters.FirstOrDefault();
            SelectedYear = AllYears.FirstOrDefault();
        }

        public IEnumerable<object> AllQuarters { get; set; }
        public IEnumerable<object> AllYears { get; set; }
        public IList<DevExpress.DemoData.Models.CustomerReport> CustomerReports { get { return NWindDataProvider.CustomerReports; } }
        public virtual object QuartersFilter { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "OnSelectedQuarterChanged")]
        public virtual object SelectedQuarter { get; set; }
        [BindableProperty(OnPropertyChangedMethodName = "OnSelectedYearChanged")]
        public virtual object SelectedYear { get; set; }
        public virtual object YearsFilter { get; set; }

        protected void OnSelectedQuarterChanged() {
            QuartersFilter = Equals(SelectedQuarter, allString) ? AllQuarters.Skip(1) : new [] { SelectedQuarter };
        }
        protected void OnSelectedYearChanged() {
            YearsFilter = new[] { SelectedYear };
        }
    }
}
